### PR TITLE
Improving testing infrastructure

### PR DESCRIPTION
**Pull request by nhibberd**
ambiata/mundane#74

### CODE REVIEW

**Comment by nhibberd** [comment]
Needs to be reworked. Add a HdfsMatcher

File: src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import com.ambiata.mundane.control._
+import com.ambiata.mundane.io._
+import com.ambiata.mundane.io.Temporary._
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import java.util.concurrent.atomic.AtomicInteger
+import scalaz._, Scalaz._
+
+case class HdfsTemporary(seed: String) {
+  private var step: AtomicInteger = new AtomicInteger(0)
```
Comment (charleso) [inline src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala]: Another `var` /supernit

**Comment by charleso** [comment]
Nick, let me know when this is ready for review.

**Comment by nhibberd** [comment]
Move TemporaryConfiguration to test tree

**Comment by nhibberd** [comment]
@charleso go for it

File: src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    else                       Hdfs.unit
+
+  def forceAddPrintFinalizer(msg: String): Hdfs[Unit] =
+    Hdfs.addFinalizer(Hdfs.putStrLn(s"Temporary: $msg"))
+}
+
+object HdfsTemporary {
+  def getTemporaryHdfsX: Hdfs[Path] = {
+    val p = new Path(uniqueDirPath.path)
+    addHdfsFinalizer(p) >> Hdfs.ok(p)
```
Comment (charleso) [inline src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala]: `addHdfsFinalizer(p).as(p)`?

Comment (nhibberd) [inline src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala]: Add a finalizer for p, so p gets cleaned up and return p. P being a Path

Comment (charleso) [inline src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala]: > Add a finalizer for p, so p gets cleaned up and return p. P being a Path

`as` is a extension method for `Functor` that's the same as doing `something.map(_ => x)`, which is what you're doing above. :)

File: src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    val incr = step.incrementAndGet.toString
+    val path = new Path(new Path(base.path, incr), seed)
+    val msg = s"Hdfs($path)"
+    addCleanupFinalizer(new Path(base.path), msg) >>
+      addPrintFinalizer(msg) >>
+        Hdfs.ok(path)
+  }
+
+  def addCleanupFinalizer(path: Path, msg: String): Hdfs[Unit] =
+    if (skipCleanup) forceAddPrintFinalizer(msg)
```
Comment (charleso) [inline src/main/scala/com/ambiata/poacher/hdfs/HdfsTemporary.scala]: Adding this here seems a little weird. Should the message be different? /notsure

File: src/test/scala/com/ambiata/poacher/hdfs/ConfigurationTemporary.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  def conf: RIO[Configuration] = for {
+    d <- LocalTemporary(path).directory
+    _ = configuration.set("hadoop.tmp.dir", d.path)
+    _ = configuration.set("scoobi.dir", d.path + "/")
+  } yield configuration
+}
+
+object ConfigurationTemporary {
+  def randomPath(conf: Configuration): ConfigurationTemporary =
```
Comment (charleso) [inline src/test/scala/com/ambiata/poacher/hdfs/ConfigurationTemporary.scala]: This seems a little awkward. At least in this PR both callsites just pass in an empty `Configuration`. And then above we're mutating the same configuration on `conf` (which then returns it). I guess I would be tempted to have the one function that creates a `Configuration`, and sets the temporary paths. (And then you can set whatever you like on it) Thoughts?

**Comment by charleso** [comment]
:hamburger:

**Pull request merged**